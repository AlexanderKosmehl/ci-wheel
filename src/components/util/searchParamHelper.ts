const urlSearchParams = new URLSearchParams(window.location.search);

export function updateSearchParams(entries: string[]) {
  urlSearchParams.set('entries', entries.join(','));
  window.history.replaceState({}, '', `${window.location.pathname}?${urlSearchParams}`);
}

export function getSearchParams() {
  return urlSearchParams.get('entries')?.split(',');
}
