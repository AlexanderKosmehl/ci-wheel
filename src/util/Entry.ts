export interface Entry {
  name: string;
  isDone: boolean;
}
