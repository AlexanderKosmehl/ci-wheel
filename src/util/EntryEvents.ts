export enum EntryEvents {
  UPDATE = 'UPDATE',
}
