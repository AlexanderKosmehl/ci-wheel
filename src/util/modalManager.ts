import generateArchiveModal from '../components/modal/archiveModal/archiveModal';
import generateImportModal from '../components/modal/importModal/importModal';
import generateSpinResultModal from '../components/modal/spinResultModal/spinResultModal';
import {
  addEntries, getCurrentEntries, toggleIsDone, updateCurrentEntries,
} from './entryManager';

let modalContainer: HTMLDivElement | undefined;

export function initModalManager() {
  const newContainerRef = document.querySelector<HTMLDivElement>('#modal-container');

  if (!newContainerRef) throw Error('No ModalContainer found!');

  modalContainer = newContainerRef;
}

const closeModal = () => {
  if (!modalContainer) return;

  modalContainer.textContent = '';
};

export function openSpinResultModal(resultLabel: string) {
  if (!modalContainer) return;

  closeModal();

  modalContainer.appendChild(
    generateSpinResultModal({
      label: resultLabel,
      onClose: () => {
        toggleIsDone(resultLabel);
        closeModal();
      },
    }),
  );
}

export function openImportModal() {
  if (!modalContainer) return;

  closeModal();

  modalContainer.appendChild(
    generateImportModal({
      onClose: closeModal,
      onImport: (importText: string) => {
        const importedEntries = importText
          .split('\n')
          .map((entry) => entry.trim())
          .filter((entry) => entry !== '');

        addEntries(importedEntries);
        closeModal();
      },
    }),
  );
}

export function openArchiveModal() {
  if (!modalContainer) return;

  closeModal();

  modalContainer.appendChild(
    generateArchiveModal({
      onClose: closeModal,
      currentEntries: getCurrentEntries().map((entry) => entry.name),
      updateCurrentEntries,
    }),
  );
}
