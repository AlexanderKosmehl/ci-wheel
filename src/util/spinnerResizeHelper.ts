function updateMaxSpinnerDimensions() {
  const root = document.querySelector<HTMLBaseElement>(':root');
  const spinnerContainer = document.querySelector('#spinner-container');
  if (!spinnerContainer || !root) return;

  root.style.setProperty(
    '--max-spinner-dimensions',
    `${Math.min(
      spinnerContainer.getBoundingClientRect().height,
      spinnerContainer.getBoundingClientRect().width,
    )}px`,
  );
}

function debounce(debouncedFunction: () => void, delay = 200) {
  let timer: NodeJS.Timeout;

  return () => {
    clearTimeout(timer);
    timer = setTimeout(() => { debouncedFunction(); }, delay);
  };
}

export default function initSpinnerResizer() {
  updateMaxSpinnerDimensions();

  const debouncedUpdate = debounce(() => updateMaxSpinnerDimensions());

  window.addEventListener('resize', debouncedUpdate);
}
