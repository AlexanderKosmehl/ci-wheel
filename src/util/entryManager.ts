import { getArchiveEntries, updateArchiveEntries } from '../components/modal/archiveModal/helper/archiveHelper';
import { Entry } from './Entry';
import { EntryEvents } from './EntryEvents';
import { getSearchParams, updateSearchParams } from './searchParamHelper';

let entries: Entry[] = [];

function dispatchUpdate() {
  window.dispatchEvent(
    new CustomEvent<Entry[]>(EntryEvents.UPDATE, { detail: entries }),
  );
}

export function initEntryManager() {
  entries = getSearchParams().map((initialValue) => ({
    name: initialValue,
    isDone: false,
  }));
  dispatchUpdate();
}

export function toggleIsDone(entryName: string) {
  entries = entries.map((entry) => {
    if (entry.name === entryName) return { name: entry.name, isDone: !entry.isDone };

    return entry;
  });

  dispatchUpdate();
}

export function getCurrentEntries() {
  return [...entries];
}

export function addEntry(newEntryName: string) {
  entries = [{ name: newEntryName, isDone: false }, ...entries];

  updateSearchParams(entries.map((entry) => entry.name));

  dispatchUpdate();
}

export function addEntries(newEntryNames: string[]) {
  entries = [
    ...newEntryNames.map((entryName) => ({ name: entryName, isDone: false })),
    ...entries,
  ];

  updateSearchParams(entries.map((entry) => entry.name));

  dispatchUpdate();
}

export function removeEntry(removedEntryName: string) {
  entries = entries.filter((entry) => entry.name !== removedEntryName);

  updateSearchParams(entries.map((entry) => entry.name));

  const archivedEntries = getArchiveEntries();
  updateArchiveEntries([removedEntryName, ...archivedEntries]);

  dispatchUpdate();
}

export function updateCurrentEntries(updatedEntryList: string[]) {
  entries = updatedEntryList.map((entryName) => ({ name: entryName, isDone: false }));

  updateSearchParams(entries.map((entry) => entry.name));

  dispatchUpdate();
}
