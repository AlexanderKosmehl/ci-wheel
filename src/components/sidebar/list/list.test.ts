import {
  describe, expect, it,
} from 'vitest';
import { addEntries } from '../../../util/entryManager';
import generateList from './list';

describe('generateListComponent', () => {
  const listComponent = generateList();
  addEntries(['Test', 'Test2']);

  it('generates component correctly', () => {
    expect(listComponent).toMatchSnapshot();
  });

  it('removes entries correctly', () => {
    const listElement = listComponent.querySelector<HTMLLIElement>('li');

    if (!listElement) throw Error('No list entries rendered!');

    listElement.querySelector<HTMLButtonElement>('[data-test=listEntryDeleteButton]')?.click();

    const remainingEntries = listComponent.querySelectorAll<HTMLSpanElement>('[data-test=listEntryLabel]');
    expect(remainingEntries.length).toBe(1);
  });
});
