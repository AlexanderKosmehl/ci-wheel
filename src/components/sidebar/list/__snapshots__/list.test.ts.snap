// Vitest Snapshot v1

exports[`generateListComponent > generates component correctly 1`] = `
<ul
  class="_listContainer_d42e02"
>
  <li
    class="_container_1761de"
  >
    <span
      class="_label_1761de"
      data-test="listEntryLabel"
    >
      Test
    </span>
    <div
      class="_buttonWrapper_1761de"
    >
      <button
        class="_button_7c1f96 _button_1761de"
        data-test="listEntryDeleteButton"
      >
        <img
          class="_icon_7c1f96"
          src="/src/assets/delete-icon.svg"
        />
      </button>
    </div>
  </li>
  <li
    class="_container_1761de"
  >
    <span
      class="_label_1761de"
      data-test="listEntryLabel"
    >
      Test2
    </span>
    <div
      class="_buttonWrapper_1761de"
    >
      <button
        class="_button_7c1f96 _button_1761de"
        data-test="listEntryDeleteButton"
      >
        <img
          class="_icon_7c1f96"
          src="/src/assets/delete-icon.svg"
        />
      </button>
    </div>
  </li>
</ul>
`;
